File: python/advanced-graphs/reconstruct-itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        adj = defaultdict(lambda: [])
        for u, v in tickets:
            adj[u].append(v)

        for location in adj:
            adj[location].sort()
            adj[location] = adj[location][::-1]

        order = []

        def dfs(curr):
            while len(adj[curr]) > 0:
                dfs(adj[curr].pop())
            order.append(curr)

        dfs("JFK")
        return order[::-1]

File: c++/advanced-graphs/reconstruct-itinerary.cc
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        std::unordered_map<std::string, std::vector<std::string>> adj;
        for (std::vector<std::string> ticket : tickets) {
            adj[ticket[0]].push_back(ticket[1]);
        }

        for (auto [from, to] : adj) {
            std::sort(adj[from].begin(), adj[from].end());
            std::reverse(adj[from].begin(), adj[from].end());
        }

        std::vector<std::string> answer;
        dfs("JFK", answer, adj);
        std::reverse(answer.begin(), answer.end());

        return answer;
    }

private:
    void dfs (std::string location, std::vector<std::string>& answer, std::unordered_map<std::string, std::vector<std::string>>& adj) {
        while (adj[location].size() > 0) {
            std::string next = adj[location].back();
            adj[location].pop_back();
            dfs(next, answer, adj);
        }
        answer.push_back(location);
    }
};
